include/pybind11/chrono.h

```diff
@@ -0,0 +1,177 @@
+/*
+    pybind11/chrono.h: Transparent conversion between std::chrono and python's datetime
+
+    Copyright (c) 2016 Trent Houliston <trent@houliston.me> and
+                       Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "pybind11.h"
+#include <cmath>
+#include <chrono>
+#include <datetime.h>
+
+NAMESPACE_BEGIN(pybind11)
+NAMESPACE_BEGIN(detail)
+
+template <typename Rep, typename Period> class type_caster<std::chrono::duration<Rep, Period>> {
+public:
+    typedef std::chrono::duration<Rep, Period> type;
+    typedef std::chrono::duration<std::chrono::hours::rep, std::ratio<86400>> days;
+
+    bool load(handle src, bool) {
+        using namespace std::chrono;
+
+        // Lazy initialise the PyDateTime import
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        if (!src) return false;
+        // If they have passed us a datetime.delta object
+        if (PyDelta_Check(src.ptr())) {
+            // The accessor macros for timedelta exist in some versions of python but not others (e.g. Mac OSX default python)
+            // Therefore we are just doing what the macros do explicitly
+            const PyDateTime_Delta* delta = reinterpret_cast<PyDateTime_Delta*>(src.ptr());
+            value = duration_cast<duration<Rep, Period>>(
+                  days(delta->days)
+                + seconds(delta->seconds)
+                + microseconds(delta->microseconds));
+            return true;
+        }
+        // If they have passed us a float we can assume it is seconds and convert
+        else if (PyFloat_Check(src.ptr())) {
+            double val = PyFloat_AsDouble(src.ptr());
+            // Multiply by the reciprocal of the ratio and round
+            value = type(std::lround(val * type::period::den / type::period::num));
```

Comment:
Two points:
- rounding is definitely not correct here for floating point `Rep` values: `std::chrono::duration<double, std::ratio<1>>` is a pretty common type with sub-second duration representation.  You definitely shouldn't be rounding when `std::chrono::treat_as_floating_point<Rep>::value` is true.
- is rounding necessary/desired here for non-floating point types?  Python itself does rounding for time deltas only at the microsecond level; the `days` value, for example, is never the rounded number of days.  If I load a python floating point value into a `std::chrono::hours` C++ object, is it really expected or desirable to get a rounded value?

The answer to both of these is probably to use `value = std::duration_cast<type>(std::chrono::duration<double>(val))`